ecosystem/typescript/sdk_v2/src/core/account.ts

```diff
@@ -0,0 +1,168 @@
+   */
+  readonly publicKey: PublicKey;
+  readonly privateKey: PrivateKey;
+
+  /**
+   * Account address associated with the account
+   */
+  readonly accountAddress: AccountAddress;
+
+  /**
+   * constructor for Account
+   *
+   * TODO: This constructor uses the nacl library directly, which only works with ed25519 keys.
+   * Need to update this to use the new crypto library if new schemes are added.
+   *
+   * @param args.privateKey PrivateKey - private key of the account
+   * @param args.address AccountAddress - address of the account
+   *
+   * This method is private because it should only be called by the factory static methods.
+   * @returns Account
+   */
+  private constructor(args: { privateKey: PrivateKey; address: AccountAddress }) {
+    const { privateKey, address } = args;
+
+    // Derive the public key from the private key
+    const keyPair = nacl.sign.keyPair.fromSeed(privateKey.toUint8Array().slice(0, 32));
+    this.publicKey = new Ed25519PublicKey({ hexInput: keyPair.publicKey });
+
+    this.privateKey = privateKey;
+    this.accountAddress = address;
+  }
+
+  /**
+   * Derives an account with random private key and address
+   *
+   * @returns Account
+   */
+  static generate(): Account {
+    const keyPair = nacl.sign.keyPair();
+    const privateKey = new Ed25519PrivateKey({ value: keyPair.secretKey.slice(0, 32) });
+    const address = new AccountAddress({ data: Account.authKey({ publicKey: keyPair.publicKey }).toUint8Array() });
+    return new Account({ privateKey, address });
+  }
+
+  /**
+   * Derives an account with provided private key
+   *
+   * @param args.privateKey Hex - private key of the account
+   * @returns Account
+   */
+  static fromPrivateKey(args: { privateKey: HexInput }): Account {
+    const privatekeyHex = Hex.fromHexInput({ hexInput: args.privateKey });
+    const keyPair = nacl.sign.keyPair.fromSeed(privatekeyHex.toUint8Array().slice(0, 32));
+    const privateKey = new Ed25519PrivateKey({ value: keyPair.secretKey.slice(0, 32) });
+    const address = new AccountAddress({ data: Account.authKey({ publicKey: keyPair.publicKey }).toUint8Array() });
+    return new Account({ privateKey, address });
+  }
+
+  /**
+   * Derives an account with provided private key and address
+   * This is intended to be used for account that has it's key rotated
+   *
+   * @param args.privateKey Hex - private key of the account
+   * @param args.address AccountAddress - address of the account
+   * @returns Account
+   */
+  static fromPrivateKeyAndAddress(args: { privateKey: HexInput; address: AccountAddress }): Account {
+    const privatekeyHex = Hex.fromHexInput({ hexInput: args.privateKey });
+    const signingKey = nacl.sign.keyPair.fromSeed(privatekeyHex.toUint8Array().slice(0, 32));
+    const privateKey = new Ed25519PrivateKey({ value: signingKey.secretKey.slice(0, 32) });
+    return new Account({ privateKey, address: args.address });
+  }
```

Comment:
These can probably be a single one with an optional address (unless we had direct feedback that it was confusing).

Additionally, this would be prime for the derive public key, and all.